### PR TITLE
[release-4.14] OCPBUGS-26573: Improve troubleshooting IC upgrades

### PR DESCRIPTION
**Pull request by ricky-rav**
If ever a cluster is not making progress during phase 1 or phase 2 of the upgrade to OVN IC, a cluster admin can now edit the interconnect configmap and add a new key (`fast-forward-to-multizone`) to bypass the two-phase upgrade and let CNO apply directly the multizone YAMLs.

The current code only allows the cluster to move forward when it's in phase 1 by setting `zone-mode=multizone` and `temporary=false`:
CurrentICupgrades drawio

Let's improve that by allowing to jump to multizone in either phase of the upgrade. At the same time, remove the `temporary` field from the configmap, since the `ongoing-upgrade` field already tracks that CNO is going through an upgrade:
ICupgrades drawio

### CODE REVIEW

File: pkg/network/ovn_kubernetes.go
Comment (tssurya) [inline pkg/network/ovn_kubernetes.go:1264]: so this variable is now removed in 4.14.7 or whatever z stream this lands in right?
how will the new upgrades from nonIC to 4.14.z where this change lands look like? I see
`remove the temporary field from the configmap, since the ongoing-upgrade field already tracks that CNO is going through an upgrade` in the PR message so that means we no longer need to know where the phase is temporary or not, just need to know if upgrade is ongoing or not? why did we need the `temporary` flag originally? seems like it was redundant with ongoingUpgrade?

I guess my TL;DR question is is there a difference between going from 4.13.0 to 4.14.0 and 4.13.0 to 4.14.z where this lands functionally for the end user if this field is going away? Do we need to make updates to the original enhancement?

File: pkg/network/ovn_kubernetes.go
Comment (tssurya) [inline pkg/network/ovn_kubernetes.go:1316]: so this is sort of a disruptive change in case ongoingUpgrade=true and someone sets this right? example, single zone is rolling out and stuck and now we set `fastForwardToMultiZone` to true to unblock and move forward with multizone.. will the single zone pods be left behind or will cleanup happen? 

Also I guess we can add something to the PR description or enhancement update around this option to say "at your own risk - to be used only with guidance from eng or something"

File: pkg/network/ovn_kubernetes.go
Comment (tssurya) [inline pkg/network/ovn_kubernetes.go:1264]: do we need to make updates to what ongoingUpgrade means? It seems like its also taking over the role that "temporary" was doing...

File: pkg/network/ovn_kubernetes.go
Comment (tssurya) [inline pkg/network/ovn_kubernetes.go:2471]: hmm maybe stupid question, but why isn't ongoing-upgrade also patched?

**Review by tssurya — COMMENTED** [review]
Overall logic is clear where you are trying to add a fast-forward knob. Left some inline questions...
Do we need some unit tests?
Did you do live testing and results were ok? (asking because we are kinda going in blind here...)

File: pkg/network/ovn_kubernetes.go
Comment (ricky-rav) [reply pkg/network/ovn_kubernetes.go:1264]: I could have simplified the logic earlier on, but I wanted to leave a window open in case we were determined to allow single zone... 

Here's how it went.

The `temporary` field was how I envisaged to distinguish between being permanently in single zone (temporary=false) or just momentarily during the upgrade to 4.14 (temporary=true). The idea was that a configmap would track an upgrade to IC or a voluntary switch to single zone.

Then I needed another flag in the configmap, `ongoing-upgrade`, so that the CNO status manager could know whether an upgrade to IC was ongoing: it would only report version=4.14 when the upgrade was done, which was signaled by `ongoingUpgrade` getting removed from the configmap.

Well after 4.14 GA, we decided not to enable switching back to single zone, so now we can finally get rid of `temporary`.

File: pkg/network/ovn_kubernetes.go
Comment (ricky-rav) [reply pkg/network/ovn_kubernetes.go:1316]: Yes, using the new flag is disruptive: CNO will push the final multizone YAMLs regardless of where the cluster is during the 4.13->4.14 upgrade. After this PR merges, my idea was to write a document about this for support. Either that or I update the enhancement...

As for the single-zone pods, they will be removed as a consequence of the new ones being pushed.

File: pkg/network/ovn_kubernetes.go
Comment (ricky-rav) [reply pkg/network/ovn_kubernetes.go:1264]: Actually, since we've decided that we're not allowing a cluster to be permanently in single zone, I might as well remove "ongoingUpgrade" and just refer to `configMapFound`, since only CNO is supposed to push the configmap now... let me try it out!

**Comment by ricky-rav** [comment]
I'll test the new changes tomorrow morning...

File: pkg/network/ovn_kubernetes.go
Comment (tssurya) [reply pkg/network/ovn_kubernetes.go]: ack did you try it? all ok on this PR?

**Review by tssurya — APPROVED** [review]
/lgtm

File: pkg/network/ovn_kubernetes.go
Comment (ricky-rav) [reply pkg/network/ovn_kubernetes.go]: yes, I retested everything after pushing the current changes: https://github.com/openshift/cluster-network-operator/pull/2076#issuecomment-1851982323

**Review by trozet — APPROVED** [review]
/approve
/label backport-risk-assessed

**Comment by ricky-rav** [comment]
@anuragthehatter for the final diagrams & configmap please refer to https://github.com/openshift/enhancements/pull/1567

**Pull request merged**
